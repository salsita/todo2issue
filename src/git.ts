import spawnAsync from '@expo/spawn-async'

export async function getCurrentBranchName (root: string) {
  const { stdout } = await spawnAsync('git', ['rev-parse', '--abbrev-ref', 'HEAD'], {
    cwd: root
  })
  return stdout.trim()
}

export interface Contribution {
  name: string
  email: string
  time: number
  timeZone: string
}

export interface BlameInfo {
  commitHash: string
  filename: string
  lineBefore: number
  lineAfter: number
  lineCount: number
  author: Contribution
  committer: Contribution
  summary: string
  text: string[]
}

const leadingLinePattern = /^([a-z0-9]{40}) ([0-9]+) ([0-9]+)( ([0-9]+))?$/
const keyValuePattern = /^([a-z\-]+) (.+)$/

const unwrapEmail = (email: string) => (/^<(.+)>$/.exec(email) ?? [undefined, email] as const)[1]

export function parseBlameInfo (blameOutput: string): BlameInfo[] {
  const results: BlameInfo[] = []
  let lastInfo: BlameInfo | null = null
  for (const line of blameOutput.split('\n')) {
    if(!lastInfo) {
      const [, commitHash, lineBefore, lineAfter, , lineCount = '1'] = leadingLinePattern.exec(line) ?? []
      if (!commitHash) {
        throw new Error(`expected leading line with hash and line numbers, got:\n\t${line}`)
      }
      lastInfo = {
        commitHash,
        lineBefore: Number(lineBefore),
        lineAfter: Number(lineAfter),
        lineCount: Number(lineCount),
        text: [],
        author: {} as Contribution,
        committer: {} as Contribution,
        filename: undefined,
        summary: ''
      }
      continue
    }

    if (lastInfo.filename != undefined) {
      lastInfo.text.push(line)
      results.push(lastInfo)
      lastInfo = null
      continue
    }

    const [, key, value] = keyValuePattern.exec(line) ?? []
    switch (key) {
      case 'author':
        lastInfo.author.name = value
        break
      case 'author-mail':
        lastInfo.author.email = unwrapEmail(value)
        break
      case 'author-time':
        lastInfo.author.time = Number(value)
        break
      case 'author-tz':
        lastInfo.author.timeZone = value
        break
      case 'committer':
        lastInfo.committer.name = unwrapEmail(value)
        break
      case 'committer-mail':
        lastInfo.committer.email = value
        break
      case 'committer-time':
        lastInfo.committer.time = Number(value)
        break
      case 'committer-tz':
        lastInfo.committer.timeZone = value
        break
      case 'summary':
        lastInfo.summary = value
        break
      case 'filename':
        lastInfo.filename = value
        break
    }
  }

  return results
}

export async function blameFile (root: string, filename: string, lines: number[]) {
  const args = [
    'blame',
    filename,
    ...lines.flatMap(line => ['-L', `${line},${line}`]),
    '--line-porcelain'
  ]
  const { stdout } = await spawnAsync('git', args, {
    cwd: root
  })
  return parseBlameInfo(stdout.trim())
}
