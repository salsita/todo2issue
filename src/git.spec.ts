import { getCurrentBranchName, parseBlameInfo } from './git'
import { resolve } from 'path'

describe('getCurrentBranchName', () => {
  it('gets something', async () => {
    const branchName = await getCurrentBranchName(resolve(__dirname, '../'))
    expect(typeof branchName).toBe('string')
    expect(branchName.includes(' ')).toBe(false)
  })
})

describe('parseBlameInfo', () => {
  it('parses blame info', async () => {
    const mockBlameOutput = `aa73a7deb88f1487c0006abcd3b684bf91c2266e 1 1 2
author Jiří Staniševský
author-mail <jirist@salsitasoft.com>
author-time 1634279203
author-tz +0200
committer Jiří Staniševský
committer-mail <jirist@salsitasoft.com>
committer-time 1634279340
committer-tz +0200
summary Add sample "code-base"
boundary
filename sample/src/index.ts
\t// TODO Single line todo comment
aa73a7deb88f1487c0006abcd3b684bf91c2266e 2 2
author Jiří Staniševský
author-mail <jirist@salsitasoft.com>
author-time 1634279203
author-tz +0200
committer Jiří Staniševský
committer-mail <jirist@salsitasoft.com>
committer-time 1634279340
committer-tz +0200
summary Add sample "code-base"
boundary
filename sample/src/index.ts
\t// TODO(#123) Single line todo comment with reference
be1ec0218067e7152167afaf1de0f84e6c55fd43 4 4 2
author Jiří Staniševský
author-mail <jirist@salsitasoft.com>
author-time 1634279489
author-tz +0200
committer Jiří Staniševský
committer-mail <jirist@salsitasoft.com>
committer-time 1634281973
committer-tz +0200
summary Scanning WiP
previous aa73a7deb88f1487c0006abcd3b684bf91c2266e sample/src/index.ts
filename sample/src/index.ts
\t// TODO(#123) Same reference different comment
be1ec0218067e7152167afaf1de0f84e6c55fd43 5 5
author Jiří Staniševský
author-mail <jirist@salsitasoft.com>
author-time 1634279489
author-tz +0200
committer Jiří Staniševský
committer-mail <jirist@salsitasoft.com>
committer-time 1634281973
committer-tz +0200
summary Scanning WiP
previous aa73a7deb88f1487c0006abcd3b684bf91c2266e sample/src/index.ts
filename sample/src/index.ts
\t// TODO Same reference different comment`

    expect(parseBlameInfo(mockBlameOutput)).toMatchSnapshot()
  })
})


