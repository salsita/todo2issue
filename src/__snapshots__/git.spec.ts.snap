// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseBlameInfo parses blame info 1`] = `
Array [
  Object {
    "author": Object {
      "email": "jirist@salsitasoft.com",
      "name": "Jiří Staniševský",
      "time": 1634279203,
      "timeZone": "+0200",
    },
    "commitHash": "aa73a7deb88f1487c0006abcd3b684bf91c2266e",
    "committer": Object {
      "email": "<jirist@salsitasoft.com>",
      "name": "Jiří Staniševský",
      "time": 1634279340,
      "timeZone": "+0200",
    },
    "filename": "sample/src/index.ts",
    "lineAfter": 1,
    "lineBefore": 1,
    "lineCount": 2,
    "summary": "Add sample \\"code-base\\"",
    "text": Array [
      "	// TODO Single line todo comment",
    ],
  },
  Object {
    "author": Object {
      "email": "jirist@salsitasoft.com",
      "name": "Jiří Staniševský",
      "time": 1634279203,
      "timeZone": "+0200",
    },
    "commitHash": "aa73a7deb88f1487c0006abcd3b684bf91c2266e",
    "committer": Object {
      "email": "<jirist@salsitasoft.com>",
      "name": "Jiří Staniševský",
      "time": 1634279340,
      "timeZone": "+0200",
    },
    "filename": "sample/src/index.ts",
    "lineAfter": 2,
    "lineBefore": 2,
    "lineCount": 1,
    "summary": "Add sample \\"code-base\\"",
    "text": Array [
      "	// TODO(#123) Single line todo comment with reference",
    ],
  },
  Object {
    "author": Object {
      "email": "jirist@salsitasoft.com",
      "name": "Jiří Staniševský",
      "time": 1634279489,
      "timeZone": "+0200",
    },
    "commitHash": "be1ec0218067e7152167afaf1de0f84e6c55fd43",
    "committer": Object {
      "email": "<jirist@salsitasoft.com>",
      "name": "Jiří Staniševský",
      "time": 1634281973,
      "timeZone": "+0200",
    },
    "filename": "sample/src/index.ts",
    "lineAfter": 4,
    "lineBefore": 4,
    "lineCount": 2,
    "summary": "Scanning WiP",
    "text": Array [
      "	// TODO(#123) Same reference different comment",
    ],
  },
  Object {
    "author": Object {
      "email": "jirist@salsitasoft.com",
      "name": "Jiří Staniševský",
      "time": 1634279489,
      "timeZone": "+0200",
    },
    "commitHash": "be1ec0218067e7152167afaf1de0f84e6c55fd43",
    "committer": Object {
      "email": "<jirist@salsitasoft.com>",
      "name": "Jiří Staniševský",
      "time": 1634281973,
      "timeZone": "+0200",
    },
    "filename": "sample/src/index.ts",
    "lineAfter": 5,
    "lineBefore": 5,
    "lineCount": 1,
    "summary": "Scanning WiP",
    "text": Array [
      "	// TODO Same reference different comment",
    ],
  },
]
`;
